jerry-ext/include/jerryscript-ext/args.h

```diff
@@ -0,0 +1,99 @@
+ */
+struct jerryx_arg_t
+{
+  jerryx_arg_transform_func_t func; /**< the transform function for the native argument */
+  void *dest; /**< point to the native argument */
+  uintptr_t extra_info; /**< extra infomation for arg transform */
+};
+
+/**
+ * The argument iteration used for transform funtions.
+ */
+struct jerryx_arg_iterator_t
+{
+  const jerryx_arg_t *c_arg_p; /**< points to the current jerryx_arg_t */
+  const jerry_value_t *js_arg_p; /**< points to the current js argument */
+  const jerry_length_t  js_arg_cnt; /**< the total count of the js arguments */
+  jerry_length_t  js_arg_index; /**< the current index of the js argument */
+};
+
+jerry_value_t jerryx_validate_and_assign_all (const jerry_value_t this_val,
+                                              const jerry_value_t *js_arg_p,
+                                              const jerry_length_t js_arg_cnt,
+                                              const jerryx_arg_t *c_arg_p,
+                                              const jerry_length_t c_arg_cnt);
+
+jerry_value_t jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p,
+                                               const jerry_length_t js_arg_cnt,
+                                               const jerryx_arg_t *c_arg_p,
+                                               const jerry_length_t c_arg_cnt);
+
+/* Helper functions for jerryx_arg_iterator_t. */
+jerry_value_t jerryx_arg_iterator_pop (jerryx_arg_iterator_t *iterator_p);
+void *jerryx_arg_iterator_get_dest (jerryx_arg_iterator_t *iterator_p);
+
+/**
+ * Indicates whether a argument can be converted into the target type.
+ */
+typedef enum
+{
+  JERRY_ARG_CONVERSION_YES, /**< the transform inside will invoke toNumber, toBoolean or toString */
+  JERRY_ARG_CONVERSION_NO /**< the conversion is not allowed. */
+} jerryx_arg_conversion_t;
+
+/**
+ * Indicates whether a argument is optional or required.
+ */
+typedef enum
+{
+  JERRY_ARG_OPTIONAL, /**< the argument is optional */
+  JERRY_ARG_REQUIRED /**< the argument is required. */
+} jerryx_arg_optional_t;
+
+/* Inline functions for initializing jerryx_arg_t */
+static inline jerryx_arg_t jerryx_arg_number (double *dest, jerryx_arg_conversion_t conv_flag, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t jerryx_arg_boolean (bool *dest, jerryx_arg_conversion_t conv_flag, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t jerryx_arg_string (char *dest, uint32_t size, jerryx_arg_conversion_t conv_flag, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t jerryx_arg_function (jerry_value_t *dest, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t jerryx_arg_native_pointer (void **dest, const jerry_object_native_info_t *info_p, jerryx_arg_optional_t opt_flag);
+static inline jerryx_arg_t jerryx_arg_ignore (void);
+static inline jerryx_arg_t jerryx_arg_custom (void *dest, uintptr_t extra_info, jerryx_arg_transform_func_t func);
+
+#include "arg-init.inline.h"
```

Comment:
Why the name is `arg-init.inline.h`:
I am not good at naming. the function inside the header are all inline functions and used to init the jerryx_arg_t.  
Do you have any suggestion about the header's name?

About the inline:
We use the inline function instead of the macro function, inline can check the input type.